### ISSUE TITLE
[Draft] Improved model specification, to clarify how multiclass classification is handled

### ISSUE THREAD

**Issue by hcho3** [dmlc/treelite#149]
The current model specification (https://github.com/dmlc/treelite/blob/3c5c3c65d024bd491efa7d5000fdffc09fa2ff2d/src/tree.proto) is not very clear about how **multi-class classification** is handled. There is no `num_class` field, so the receiving application needs to infer the number of classes from `num_output_group` as well as `random_forest_flag`. I am proposing a new model spec to explicitly represent multiple methods for multi-class classification.

Methods for performing multi-class classification
=====

Here, we discuss multiple methods for performing multi-class classification with a decision tree ensemble. The proposed spec should accommodate all the methods discussed.

**Method 1: Gradient boosting**. We train `[number of iterations] * [number of classes]` trees, where each leaf node produces a single real-valued output. To compute the class probabilities, we gather outputs from all trees and assign them to multiple "output groups" in round-robin fashion. There should be `[number of classes]` output groups. Then we compute the softmax over the partial sums. Let us look at a concrete example to see what's going on.

Let's say we use XGBoost to solve the classification problem with the [Iris dataset](https://archive.ics.uci.edu/ml/datasets/Iris). The label in Iris consists of 3 distinct classes. Running XGBoost for 4 boosting iterations results in a 12-tree ensemble. Now given an input `x`, how do we obtain the class probabilities (3 probabilities, one for each class)? Say we got the following outputs from the 12 trees:

- Tree 0 produces +0.5
- Tree 1 produces +1.5
- Tree 2 produces -2.3
- Tree 3 produces -1.5
- Tree 4 produces +0.1
- Tree 5 produces +0.7
- Tree 6 produces +1.5
- Tree 7 produces -0.9
- Tree 8 produces -0.1
- Tree 9 produces +0.3
- Tree 10 produces -0.7
- Tree 11 produces +0.2

Since there are 3 classes, we create 3 output groups. We assign each output to an output group with this rule: `[Group ID] = [Tree ID] % [number of classes]`.

- Output group 0 gets Trees 0, 3, 6, and 9, and the sum for this output group is (+0.5) + (-1.5) + (+1.5) + (+0.3) = +0.8
- Output group 1 gets Trees 1, 4, 7, and 10, and the sum for this output group is (+1.5) + (+0.1) + (-0.9) + (-0.7) = 0.0
- Output group 2 gets Trees 2, 5, 8, and 11, and the sum for this output group is (-2.3) + (+0.7) + (-0.1) + (+0.2) = -1.5

The vector `[+0.8, 0.0, -1.5]` consisting of the three subtotals quantifies the relative likelihood of the label classes. Since the first element (+0.8) is the largest, the first class is the most likely outcome for the particular data point. This vector is not yet a probability distribution, since its elements do not sum to 1. We apply the softmax function to get a proper probability distribution:

```pycon
>>> import numpy as np
>>> x = np.array([+0.8, 0.0, -1.5])
>>> expx = np.exp(x)
>>> expx / expx.sum()
array([0.6453329 , 0.28996677, 0.06470033])
```
We interpret the result as giving 64.5% probability for the first class, 29.0% for the second, and 6.5% for the third.

Method 1 is used by XGBoost, LightGBM, and `GradientBoostingClassifier` from Scikit-Learn.

**Method 2: Random Forest**. Each leaf node produces a hard class prediction. To compute the class probabilities, we gather outputs from all trees and count the votes made for each class.

Let's say we use `RandomForestClassifier` from Scikit-Learn to classify the Iris dataset and we set `n_estimators` to 12. A typical outcome would be as follows:

- Tree 0 produces 0
- Tree 1 produces 1
- Tree 2 produces 2
- Tree 3 produces 1
- Tree 4 produces 2
- Tree 5 produces 1
- Tree 6 produces 2
- Tree 7 produces 2
- Tree 8 produces 1
- Tree 9 produces 2
- Tree 10 produces 0
- Tree 11 produces 1

We count 2 votes for the first class, 6 votes for the second class, and 4 votes for the third class. We obtain class probabilities by dividing the vote counts by the total number of trees:
```pycon
>>> import numpy as np
>>> x = np.array([2, 6, 4])
>>> x / x.sum()
array([0.16666667, 0.5       , 0.33333333])
```

Method 2 is used by `RandomForestClassifier` from Scikit-Learn.

**Method 3**. Each leaf node produces an one-hot-encoded vector encoding a hard class prediction. For example, 12 trees may produce the following:

- Tree 0 produces `[1, 0, 0]`
- Tree 1 produces `[0, 1, 0]`
- Tree 2 produces `[0, 1, 0]`
- Tree 3 produces `[0, 1, 0]`
- Tree 4 produces `[0, 0, 1]`
- Tree 5 produces `[0, 1, 0]`
- Tree 6 produces `[0, 0, 1]`
- Tree 7 produces `[0, 0, 1]`
- Tree 8 produces `[0, 1, 0]`
- Tree 9 produces `[0, 0, 1]`
- Tree 10 produces `[1, 0, 0]`
- Tree 11 produces `[0, 1, 0]`

Consider the sum of these 12 vectors: `[2, 6, 4]`. Note that this sum is simply the number of votes given to each class (2 for the first class, 6 for the second, 4 for the third). Thus, the average of the 12 vectors gives the class probabilities.

I am not aware of any decision tree library that uses Method 3. However, currently Treelite uses Method 3 internally, since Treelite does not have a facility to conduct voting.

Proposed model specification
============
To make explicit which Method is being used, the proposed model spec will contain a new enum field `multiclass_mode`. We also add `num_class` field to explicitly indicate how many label classes there are:
```protobuf
enum MultiClassMode {
  NONE = 0;                      // Not multiclass classifier
  SOFTMAX_WITH_REAL_OUTPUT = 1;  // Method 1 and 3
  VOTING_WITH_CLASS_OUTPUT = 2;  // Method 2
}

message Model {
  ...
  optional MultiClassMode multiclass_node = 6;
  optional int32 num_class = 7;
  ...
}
```
Furthermore, we allow multiple data types for leaf outputs:
```protobuf

// Inspired by TensorFlow SavedModel format
enum DataType {
  DT_INVALID = 0;
  DT_FLOAT = 1;
  DT_DOUBLE = 2;
  DT_FLOAT16 = 3;
  DT_BFLOAT16 = 4;
  DT_INT8 = 5;
  DT_UINT8 = 6;
  DT_INT16 = 7;
  DT_UINT16 = 8;
  DT_INT32 = 9;
  DT_UINT32 = 10;
  DT_INT64 = 11;
  DT_UINT64 = 12;
}

message Value {
  optional DataType dtype = 1;
  // DT_FLOAT
  optional float float_val = 2 [packed = true];
  // DT_DOUBLE
  optional double double_val = 3 [packed = true];
  // DT_INT8, DT_INT16, DT_INT32
  optional int32 int_val = 4 [packed = true];
  // DT_INT64
  optional int64 int64_val = 5 [packed = true];
  // DT_UINT8, DT_UINT16, DT_UINT32
  optional uint32 uint_val = 6 [packed = true];
  // DT_UINT64
  optional uint64 uint64_val = 7 [packed = true];
  // DT_FLOAT16, DT_BFLOAT16
  optional bytes float16_val = 8;
}

message Node {
  ...
  optional Value leaf_value = 9;
  repeated Value leaf_vector = 10; 
  ...
}
```
Note that 8-bit and 16-bit integers will be promoted to 32-bit before they are saved to `int_val` or `uint_val` field. This is because Protobuf does not provide 8-bit or 16-bit integer type. For a similar reason, we use `bytes` to store 16-bit floating point type. Little endian shall be used when serializing 16-bit floating point to bytes.

When `multiclass_mode` is set to `SOFTMAX_WITH_REAL_OUTPUT`, each leaf node may either produce a single output or a vector of any type. On the other hand, when `multiclass_mode` is `VOTING_WITH_CLASS_OUTPUT`, each leaf must produce a single output and that output must be of integer type.

@canonizer @levsnv Please review. Feedback would be appreciated.

cc @JohnZed @trivialfis

**Comment by hcho3** [comment]
For now, I didn't include `bfloat16` in the data type enum. Maybe we should. To exactly represent `bfloat16`, we'll need to use `bytes` type.

**Comment by levsnv** [comment]
Thank you for starting the discussion!

Several questions:
1. What's the reason to allow leaves to have different dtype in `leaf_value`?. Thinking of deserialization times and disk/network footprint.
2. Same WRT `leaf_vector` - this even seems to allow a vector out of heterogeneous dtypes.
3.1.1. For models where trees predict vectors, is there a way to specify no softmax/another epilogue?
3.1.2. Does this support vector regression? If so, which `MultiClassMode` would it use?
3.2. Would you allow some leaves in a tree to produce a `leaf_value` while others produce a `leaf_vector`? If so, how would you interpret the `leaf_value`?
4. Say, all leaves produce vectors. Does that automatically imply Method 3 (and not Method 1)?
5. One could imagine two possible ways to convert probabilities into predictions:
5.a) apply threshold
5.b) argmax
The difference being potential applications where in a tie, it's better to predict no label (empty label) as opposed to a random label.
Is argmax not used anywhere except cuML RF?
Would one be able to save this choice within the treelite model?

**Comment by hcho3** [comment]
@levsnv Let's tackle the first question. Currently, LightGBM and scikit-learn uses 64-bit floating point for thresholds and leaf outputs, whereas XGBoost uses 32-bit floating point.  So to produce accurate results, we need a way to specify the data type of thresholds. There are many open issues around LightGBM models getting incorrect results: #77 #94 #155. So if Treelite were to claim to support tree libraries other than XGBoost, it needs to allow users to specify the type of threshold and leaf outputs.

I do realize that a type-general spec is not necessarily performant. [In my latest PoC](https://github.com/hcho3/treelite/tree/poc_faster_serializer), I use float32 exclusively, to make serialization efficient. Also, I found Protobuf to be rather slow; dumping binary is faster.

A way forward is to distinguish a **specification** and **serialization format**.
* **Spec** should be sufficiently general to capture all tree models, and it is to be expressed in Protobuf.
* **Serialization format** can be specialized to a subset of models, e.g. trees with float32 thresholds. Serialization format is a lightweight representation of C++ objects, to be found in `include/xgboost/tree.h`. 

As for 2: It is not my intent to allow heterogenous types.

**Comment by hcho3** [comment]
@levsnv For now, please evaluate whether the spec contains sufficient information to perform multi-class classification. Let's discuss about fast serialization in #160.

**Comment by hcho3** [comment]
> 3.1.1. For models where trees predict vectors, is there a way to specify no softmax/another epilogue?

Yes, by specifying the transform method (`pred_transform`) in `extra_params` field.  You'd set `pred_transform='identity_multiclass'` to indicate that there should be no softmax. See the list of possible values at https://treelite.readthedocs.io/en/latest/knobs/model_param.html.

> 3.1.2. Does this support vector regression? If so, which MultiClassMode would it use?

Yes, just set it to `NONE`.

> 3.2. Would you allow some leaves in a tree to produce a leaf_value while others produce a leaf_vector? If so, how would you interpret the leaf_value?

No, that would not make sense. Currently, Treelite will throw an error if such model is specified.

> 3. Say, all leaves produce vectors. Does that automatically imply Method 3 (and not Method 1)?

Yes, since Method 1 currently assumes that each leaf produces `leaf_value`.

> 4. One could imagine two possible ways to convert probabilities into predictions ...

Currently, you can set `pred_transform` parameter to `softmax` or `max_index`.

**Comment by levsnv** [comment]
Re: 1. Judging by the numbers in #156, indeed, this will not be the best serialization format.
However, you could limit the spec like this https://gist.github.com/levsnv/7028c04aa68dee14852e932e6568fdd0 to avoid having to code checks for those edge cases (which, I assume, you didn't plan on supporting?)

**Comment by levsnv** [comment]
For integer output when the threshold is a different type, can we restrict the integer to int32? Should be plenty of classes, and simplifies code.

**Comment by hcho3** [comment]
Sounds good to me.

**Comment by levsnv** [comment]
Do you plan to support using a single model for https://scikit-learn.org/0.22/modules/ensemble.html#bagging-meta-estimator ? Or would the user need to do the averaging themselves in this case? Not aware of any (potential) implementations yet outside sklearn, just asking for clarity.

**Comment by levsnv** [comment]
Overall, I personally don't see any risks in this format.
Pinging re: question above, otherwise, signing off :) Of course, we'd be remiss without @canonizer's review

**Comment by hcho3** [comment]
@levsnv I don't think we'll cover bagging meta estimator. We already have plenty of things to do with trees.

**Comment by canonizer** [comment]
1. Do we really need the extra `MultiClassMode`? It seems to me that what it intends to do is already covered by the `pred_transform`. In case integer class labels are stored in leaves, they may be interpreted as if they were the corresponding one-hot vectors for the purpose of how they contribute to the final result. Of course, it is still possible to use specific optimizations for this case.

2. Maybe we'll need to have a field that indicates that integers are stored in leaves. Also, are you planning to have an interpretation of integers in leaves different from the one described above?

2. I agree that having the explicit `num_class` field is useful. The name is much clearer than `num_output_group`.

3. Regarding storing integers in leaves, int32 should be enough at first. Of course, feel free to support other integer types.

**Comment by hcho3** [comment]
@canonizer 

1. Hmm I suppose you are right. If the leaf output is an integer, the only reasonable interpretation is that it's a class output.

2. 

> we'll need to have a field that indicates that integers are stored in leaves

Since Protobuf doesn't support type generics, we'll need to define separate message types for each type combination, e.g.
```
Protobuf message -> C++ class (treelite::tree<ThresholdType, LeafType>
TreeFloat32 -> treelite::Tree<float32, float32>
TreeFloat64 -> treelite::Tree<float64, float64>
TreeFloat32IntOut -> treelite::Tree<float32, int32>
TreeFloat64IntOut -> treelite::Tree<float64, int32>
```

> Also, are you planning to have an interpretation of integers in leaves different from the one described above?

No. See 1.

**Comment by levsnv** [comment]
@hcho we could just use `has_float_forest_intleaf()` etc. to determine what the leaves store (see gist above)? To avoid duplicate labels and related edge cases.
